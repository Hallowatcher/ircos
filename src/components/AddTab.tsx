
import * as React from 'react';

interface IProps {
  clickAddTab: () => void;
}

const divStyle: React.CSSProperties = {
  padding: '10px',
  margin: '0 2px',
  display: 'flex',
  height: 25,
  alignItems: 'center',
  cursor: 'pointer',
  backgroundColor: 'rgb(241, 196, 15)',
  fontWeight: 800,
  color: 'white',
  boxSizing: 'border-box',
  WebkitUserSelect: 'none'
};

export class AddTab extends React.Component<IProps, any> {
  public render() {
    return (
      <div style={divStyle} onClick={this.props.clickAddTab}>+</div>
    );
  }
}
