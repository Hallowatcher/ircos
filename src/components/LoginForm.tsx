import * as React from 'react';

type Props = {
  submitLogin: (user: string, pass: string) => void
}

type State = {
  user: string,
  pass: string
}

export class LoginForm extends React.Component<Props, State> {

  constructor(props: any) {
    super(props);

    this.state = {
      user: '',
      pass: ''
    }
  }

  inputChanged(event: any) {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleSubmit(event: any) {
    event.preventDefault();

    if (this.state.user !== '' && this.state.pass !== '')
      this.props.submitLogin(this.state.user, this.state.pass);
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit.bind(this)}>
        <input type="text" name="user" value={this.state.user} onChange={this.inputChanged.bind(this)} />
        <input type="password" name="pass" value={this.state.pass} onChange={this.inputChanged.bind(this)} />
        <input type="submit" />
      </form>
    )
  }
}