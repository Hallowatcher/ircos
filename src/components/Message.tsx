import * as React from 'react';
import * as moment from 'moment';
import { IMessage, MessageType } from '../models';

interface IProps {
  nick: string;
  text: string;
  type: MessageType;
  date: moment.Moment;

  userClicked: (user: string) => void;
}

interface IStyles {
  [key: string]: {
    [key: string]: React.CSSProperties
  } | React.CSSProperties;
}

const styles: IStyles = {
  users: {
    regular: {
      color: '#ffc561',
      fontWeight: 800,
      cursor: 'pointer'
    },
    moderator: {
      color: '#BF3434',
      fontWeight: 800,
      cursor: 'pointer'
    },
    self: {
      color: '#CECECE',
      fontWeight: 800,
      cursor: 'pointer'
    }
  },
  message: {},
  systemMessage: {
    color: '#ADADAD'
  },
  time: {
    color: '#b2b2b2'
  }
};

export class Message extends React.Component<IProps, any> {

  public render() {

    let userStyle = styles.users.regular;
    const { nick, text, date, type } = this.props;
    switch (type) {
      case MessageType.self:
        userStyle = styles.users.self;
        break;
    }

    const momentDate = moment(date);

    // Build date
    const hours = ('0' + momentDate.hours()).slice(-2);
    const minutes = ('0' + momentDate.minutes()).slice(-2);

    let message = (
      <div>
        <span style={styles.time}>[{hours}:{minutes}] </span>
        <span style={userStyle} onClick={this.userClicked}>{nick}: </span>
        <span style={styles.message}>{text}</span>
      </div>
    );

    if (type === MessageType.system) {
      message = (
        <div>
          <span style={styles.systemMessage}>{text}</span>
        </div>
      );
    }

    if (type === MessageType.action) {
      message = (
        <div>
          <span style={styles.time}>[{hours}:{minutes}] </span>
          <span style={styles.message} onClick={this.userClicked}>* {nick} </span>
          <span style={styles.message}>{text}</span>
        </div>
      );
    }

    return message;
  }

  private userClicked = () => {
    this.props.userClicked(this.props.nick);
  }
}
