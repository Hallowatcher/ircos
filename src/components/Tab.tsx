
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {
  DragSource,
  DropTarget,
  DropTargetSpec,
  DragSourceSpec,
  DragSourceCollector,
  DropTargetCollector
} from 'react-dnd';
import { flow } from 'lodash';

interface IProps {
  index: number;
  tabName: string;
  tabClick: (name: string) => void;
  tabMove: (from: number, to: number) => void;
  closeTab: (name: string) => void;
  isActive: boolean;

  // Drag
  connectDragSource?: any;
  connectDropTarget?: any;
  isDragging?: boolean;
  isOver?: boolean;
}

const xStyle: React.CSSProperties = {
  padding: '0 10px',
  margin: '0 0 0 10px',
  backgroundColor: 'rgba(255, 255, 255, 0.15)',
  height: '100%',
  display: 'flex',
  alignItems: 'center',
  fontWeight: 400
};

const divStyle: React.CSSProperties = {
  padding: '0 0 0 10px',
  margin: '0 2px',
  display: 'flex',
  height: 25,
  alignItems: 'center',
  cursor: 'pointer',
  backgroundColor: 'rgba(255, 255, 255, 0.15)',
  fontWeight: 800,
  color: 'white',
  boxSizing: 'border-box',
  WebkitUserSelect: 'none'
};

const divStyleActive: React.CSSProperties = {
  ...divStyle,
  backgroundColor: 'rgba(255, 255, 255, 0.3)'
};

const divStyleDragging: React.CSSProperties = {
  outline: 'white dashed 1px'
};

const tabSource: DragSourceSpec<IProps> = {
  beginDrag(props) {
    return {
      name: props.tabName,
      index: props.index
    };
  }
};

const tabTarget: DropTargetSpec<IProps> = {
  hover(props, monitor, component) {
    const dragIndex = (monitor.getItem() as any).index;
    const hoverIndex = props.index;

    // Don't replace items with themselves
    if (dragIndex === hoverIndex) {
      return;
    }

    // Determine rectangle on screen
    const hoverBoundingRect = ReactDOM.findDOMNode(component).getBoundingClientRect();

    // Get horizontal middle
    const hoverMiddleX = (hoverBoundingRect.right - hoverBoundingRect.left) / 2;

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    // Get pixels to the left
    const hoverClientX = clientOffset.x - hoverBoundingRect.left;

    // Only perform the move when the mouse has crossed half of the items height
    // When dragging downwards, only move when the cursor is below 50%
    // When dragging upwards, only move when the cursor is above 50%

    // Dragging left
    if (dragIndex < hoverIndex && hoverClientX < hoverMiddleX) {
      return;
    }

    // Dragging right
    if (dragIndex > hoverIndex && hoverClientX > hoverMiddleX) {
      return;
    }

    // Time to actually perform the action
    props.tabMove(dragIndex, hoverIndex);

    // Note: we're mutating the monitor item here!
    // Generally it's better to avoid mutations,
    // but it's good here for the sake of performance
    // to avoid expensive index searches.
    (monitor.getItem() as any).index = hoverIndex;
  }
};

let collectDrag: DragSourceCollector;
collectDrag = function(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  };
};

let collectDrop: DropTargetCollector;
collectDrop = function(connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver()
  };
};

class Tab extends React.Component<IProps, null> {
  constructor() {
    super();

    this.handleClick = this.handleClick.bind(this);
    this.handleMiddleClick = this.handleMiddleClick.bind(this);
  }

  public render() {
    const { connectDragSource, connectDropTarget, isDragging, isActive } = this.props;
    return connectDragSource(connectDropTarget(
      <div
        onMouseDown={this.handleMiddleClick}
        style={{...(isDragging ? divStyleDragging : {}), ...(isActive ? divStyleActive : divStyle)}}
      >
        <div
          onClick={this.handleClick}
        >{this.props.tabName}
        </div>
        <div onClick={this.props.closeTab.bind(this, this.props.tabName)} style={xStyle}>&times;</div>
      </div>
    ));
  }
  private handleMiddleClick(event) {
    const type = event.nativeEvent.which;
    if (type !== 2) { return; }
    this.props.closeTab(this.props.tabName);
  }

  private handleClick(event) {
    this.props.tabClick(this.props.tabName);
  }
}

export default flow(
  DragSource('tab', tabSource, collectDrag),
  DropTarget('tab', tabTarget, collectDrop)
)(Tab);
