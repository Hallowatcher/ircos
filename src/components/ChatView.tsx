
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Message } from './Message';
import * as moment from 'moment';
import { IMessage } from '../models';

interface IProps {
  messages: any[];
  selfNick: string;
  userClicked: (user: string) => void;
  style?: React.CSSProperties;
}

export class ChatView extends React.Component<IProps, any> {

  private scrollAtBottom: boolean = true;
  private scrollAtTop: boolean = false;
  private topMessage: any = null;
  private historyChanged: boolean = false;
  private multiplier: number = 1;

  public componentWillUpdate(nextProps: IProps) {
    this.historyChanged = nextProps.messages.length !== this.props.messages.length;
    if (!this.historyChanged || !this.refs.container) { return; }

    const container = this.refs.container as HTMLDivElement;
    const scrollPos = Math.ceil(container.scrollTop);
    const scrollBottom = Math.ceil(container.scrollHeight - container.clientHeight);
    this.scrollAtBottom = (scrollBottom <= 0) || (scrollPos === scrollBottom);
  }

  public componentDidUpdate() {
    if (this.scrollAtBottom) {
      this.scrollToBottom();
    }
  }

  public render() {
    return (
      <div style={this.props.style} ref="container">
        {this.mapMessages()}
      </div>
    );
  }

  /* istanbul ignore next */
  private scrollToBottom() {
    const container = this.refs.container as HTMLDivElement;
    const scrollHeight = container.scrollHeight;
    const height = container.clientHeight;
    const maxScrollTop = scrollHeight - height;
    ReactDOM.findDOMNode(container).scrollTop = maxScrollTop > 0 ? maxScrollTop : 0;
  }

  private mapMessages() {
    return this.props.messages.map((msg: IMessage, index: number) => {
        return (
          <Message
            key={index}
            nick={msg.nick}
            text={msg.text}
            type={msg.type}
            date={msg.date}
            userClicked={this.props.userClicked}
          />
        );
      });
  }
}
