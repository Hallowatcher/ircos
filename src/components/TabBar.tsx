
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import Tab from './Tab';
import { AddTab } from './AddTab';
import * as FlipMove from 'react-flip-move';

interface IProps {
  tabs: string[];
  currentChannel: string;

  tabClick: (name: string) => void;
  tabMove: (from: number, to: number) => void;
  tabClose: (name: string) => void;
  tabAdd: () => void;
}

const style: React.CSSProperties = {
  flexGrow: 1,
  padding: '0 5px',
  display: 'flex',
  overflowY: 'hidden',
  overflowX: 'auto',
  alignItems: 'flex-end'
};

export class TabBar extends React.Component<IProps, any> {

  public render() {
    const channelMap = this.props.tabs.map((channel, index) =>
    (
      <Tab
        key={channel}
        ref={channel}
        index={index}
        tabName={channel}
        tabClick={this.props.tabClick}
        tabMove={this.props.tabMove}
        closeTab={this.props.tabClose}
        isActive={this.props.currentChannel === channel}
      />
    )
    );

    return (
      <FlipMove style={style} duration={125} easing="ease-out">
        {channelMap}
        <AddTab clickAddTab={this.props.tabAdd} key="add" ref="add" />
      </FlipMove>
    );
  }
}
