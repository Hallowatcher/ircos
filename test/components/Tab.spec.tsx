
import * as React from 'react';
import { Tab } from '../../src/components/Tab';
import * as enzyme from 'enzyme';

describe('AddTab', function() {

  let component: enzyme.ShallowWrapper<any, any>;
  let tabName;
  let tabClick;
  let closeTab;
  let isActive;
  let preventDefault;

  beforeEach(function() {

    // Setup
    tabName = 'foo123';
    tabClick = jest.fn();
    closeTab = jest.fn();
    preventDefault = jest.fn();
    isActive = true;

    component = enzyme.shallow(
      <Tab tabName={tabName} tabClick={tabClick} closeTab={closeTab} isActive={isActive} />
    );
  });

  it('should render', function() {
    expect(component).toMatchSnapshot();
  });

  it('should have clickable name div', function() {
    const name = component.find('div > div').first();
    expect(name.text()).toEqual(tabName);
    expect(name.props().onClick).toBeDefined();
  });

  it('should have clickable x to close', function() {
    const x = component.find('div > div').last();
    expect(x.text()).toEqual('Ã—');
    expect(x.props().onClick).toBeDefined();
  });

  it('should render with inactive class', function() {
    component.setProps({ isActive: false });
    expect(component).toMatchSnapshot();
  });

  it('should call close handler', function() {
    component.find('div > div').last().simulate('click', { preventDefault });
    expect(closeTab).toHaveBeenCalledWith(tabName, { preventDefault });
  });

  it('should call click handler', function() {
    component.find('div > div').first().simulate('click', { preventDefault });
    expect(tabClick).toHaveBeenCalled();
  });

  it('should close tab on middle click handler', function() {
    component.find('div').first().simulate('mousedown', { preventDefault, nativeEvent: { which: 2 } });
    expect(closeTab).toHaveBeenCalled();
  });

  it('should not close tab', function() {
    component.find('div').first().simulate('mousedown', { preventDefault, nativeEvent: { which: 1 } });
    expect(closeTab).toHaveBeenCalledTimes(0);
  });
});
