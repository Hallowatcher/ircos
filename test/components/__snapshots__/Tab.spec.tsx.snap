// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab should render 1`] = `
<DragDropContext(DropTarget(DragSource(Tab)))
  closeTab={[Function]}
  index={1}
  isActive={true}
  isDragging={false}
  tabClick={[Function]}
  tabMove={[Function]}
  tabName="foo123"
>
  <DropTarget(DragSource(Tab))
    closeTab={[Function]}
    index={1}
    isActive={true}
    isDragging={false}
    tabClick={[Function]}
    tabMove={[Function]}
    tabName="foo123"
  >
    <DragSource(Tab)
      closeTab={[Function]}
      connectDropTarget={[Function]}
      index={1}
      isActive={true}
      isDragging={false}
      isOver={false}
      tabClick={[Function]}
      tabMove={[Function]}
      tabName="foo123"
    >
      <Tab
        closeTab={[Function]}
        connectDragSource={[Function]}
        connectDropTarget={[Function]}
        index={1}
        isActive={true}
        isDragging={false}
        isOver={false}
        tabClick={[Function]}
        tabMove={[Function]}
        tabName="foo123"
      >
        <div
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "WebkitUserSelect": "none",
              "alignItems": "center",
              "backgroundColor": "rgba(255, 255, 255, 0.3)",
              "boxSizing": "border-box",
              "color": "white",
              "cursor": "pointer",
              "display": "flex",
              "fontWeight": 800,
              "height": 25,
              "margin": "0 2px",
              "padding": "0 0 0 10px",
            }
          }
        >
          <div>
            foo123
          </div>
          <div
            onClick={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.15)",
                "display": "flex",
                "fontWeight": 400,
                "height": "100%",
                "margin": "0 0 0 10px",
                "padding": "0 10px",
              }
            }
          >
            ×
          </div>
        </div>
      </Tab>
    </DragSource(Tab)>
  </DropTarget(DragSource(Tab))>
</DragDropContext(DropTarget(DragSource(Tab)))>
`;

exports[`Tab should render with inactive class 1`] = `
<DragDropContext(DropTarget(DragSource(Tab)))
  closeTab={[Function]}
  index={1}
  isActive={false}
  isDragging={false}
  tabClick={[Function]}
  tabMove={[Function]}
  tabName="foo123"
>
  <DropTarget(DragSource(Tab))
    closeTab={[Function]}
    index={1}
    isActive={false}
    isDragging={false}
    tabClick={[Function]}
    tabMove={[Function]}
    tabName="foo123"
  >
    <DragSource(Tab)
      closeTab={[Function]}
      connectDropTarget={[Function]}
      index={1}
      isActive={false}
      isDragging={false}
      isOver={false}
      tabClick={[Function]}
      tabMove={[Function]}
      tabName="foo123"
    >
      <Tab
        closeTab={[Function]}
        connectDragSource={[Function]}
        connectDropTarget={[Function]}
        index={1}
        isActive={false}
        isDragging={false}
        isOver={false}
        tabClick={[Function]}
        tabMove={[Function]}
        tabName="foo123"
      >
        <div
          onClick={[Function]}
          onMouseDown={[Function]}
          style={
            Object {
              "WebkitUserSelect": "none",
              "alignItems": "center",
              "backgroundColor": "rgba(255, 255, 255, 0.15)",
              "boxSizing": "border-box",
              "color": "white",
              "cursor": "pointer",
              "display": "flex",
              "fontWeight": 800,
              "height": 25,
              "margin": "0 2px",
              "padding": "0 0 0 10px",
            }
          }
        >
          <div>
            foo123
          </div>
          <div
            onClick={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(255, 255, 255, 0.15)",
                "display": "flex",
                "fontWeight": 400,
                "height": "100%",
                "margin": "0 0 0 10px",
                "padding": "0 10px",
              }
            }
          >
            ×
          </div>
        </div>
      </Tab>
    </DragSource(Tab)>
  </DropTarget(DragSource(Tab))>
</DragDropContext(DropTarget(DragSource(Tab)))>
`;
